spring-kafka/src/main/java/org/springframework/kafka/listener/ConsumerProperties.java

```diff
@@ -0,0 +1,325 @@
+
+import org.springframework.kafka.support.LogIfLevelEnabled;
+import org.springframework.kafka.support.TopicPartitionOffset;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Common consumer properties.
+ *
+ * @author Gary Russell
+ * @since 2.3
+ *
+ */
+public class ConsumerProperties {
+
+	/**
+	 * The default {@link #setPollTimeout(long) pollTimeout} (ms).
+	 */
+	public static final long DEFAULT_POLL_TIMEOUT = 5_000L;
+
+	/**
+	 * Topic names.
+	 */
+	private final String[] topics;
+
+	/**
+	 * Topic pattern.
+	 */
+	private final Pattern topicPattern;
+
+	/**
+	 * Topics/partitions/initial offsets.
+	 */
+	private final TopicPartitionOffset[] topicPartitionsToAssign;
+
+	/**
+	 * The max time to block in the consumer waiting for records.
+	 */
+	private long pollTimeout = DEFAULT_POLL_TIMEOUT;
+
+	/**
+	 * Override the group id.
+	 */
+	private String groupId;
+
+	/**
+	 * Override the client id.
+	 */
+	private String clientId = "";
+
+	/**
+	 * A user defined {@link ConsumerRebalanceListener} implementation.
+	 */
+	private ConsumerRebalanceListener consumerRebalanceListener;
+
+	private Duration syncCommitTimeout;
+
+	/**
+	 * The commit callback; by default a simple logging callback is used to log
+	 * success at DEBUG level and failures at ERROR level.
+	 */
+	private OffsetCommitCallback commitCallback;
+
+	/**
+	 * Whether or not to call consumer.commitSync() or commitAsync() when the
+	 * container is responsible for commits. Default true.
+	 */
+	private boolean syncCommits = true;
+
+	private LogIfLevelEnabled.Level commitLogLevel = LogIfLevelEnabled.Level.DEBUG;
+
+	/**
+	 * Create properties for a container that will subscribe to the specified topics.
+	 * @param topics the topics.
+	 */
+	public ConsumerProperties(String... topics) {
+		Assert.notEmpty(topics, "An array of topics must be provided");
+		this.topics = topics.clone();
+		this.topicPattern = null;
+		this.topicPartitionsToAssign = null;
+	}
+
+	/**
+	 * Create properties for a container that will subscribe to topics matching the
+	 * specified pattern. The framework will create a container that subscribes to all
+	 * topics matching the specified pattern to get dynamically assigned partitions. The
+	 * pattern matching will be performed periodically against topics existing at the time
+	 * of check.
+	 * @param topicPattern the pattern.
+	 * @see org.apache.kafka.clients.CommonClientConfigs#METADATA_MAX_AGE_CONFIG
+	 */
+	public ConsumerProperties(Pattern topicPattern) {
+		this.topics = null;
+		this.topicPattern = topicPattern;
+		this.topicPartitionsToAssign = null;
+	}
+
+	/**
+	 * Create properties for a container that will assign itself the provided topic
+	 * partitions.
+	 * @param topicPartitions the topic partitions.
+	 * @deprecated in favor of {@link #ContainerProperties(TopicPartitionOffset...)}.
+	 */
+	@Deprecated
```

Comment:
This class is new and I believe it really can be set free from any deprecated code.

Reply:
Right; that's why I made it package protected - I can't do the assertion in the subclass before calling `super(...)`.

Reply:
But there's another way...